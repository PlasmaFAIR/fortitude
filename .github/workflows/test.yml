name: Tests

on:
  push:
    paths:
      - '**.rs'
      - '**/Cargo.toml'
      - '**/Cargo.lock'
      - '.github/workflows/test.yml'
  pull_request:
    paths:
      - '**.rs'
      - '**/Cargo.toml'
      - '**/Cargo.lock'
      - '.github/workflows/test.yml'
  workflow_dispatch:

env:
  CARGO_INCREMENTAL: 0
  CARGO_TERM_COLOR: always

jobs:

  build-linux:
    name: Build and Test
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: "Install cargo nextest"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-nextest
    - name: "Install cargo insta"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-insta
    - name: "Build"
      run: cargo build --verbose
    - name: "Run tests"
      shell: bash
      env:
        NEXTEST_PROFILE: "ci"
      run: cargo insta test --all-features --unreferenced reject --test-runner nextest
    - name: "Check docs"
      run: cargo dev generate-all --mode=check
    - name: "Check docs for broken links"
      run: cargo doc --all --no-deps
      env:
        # Setting RUSTDOCFLAGS because `cargo doc --check` isn't yet
        # implemented (https://github.com/rust-lang/cargo/issues/10025)
        RUSTDOCFLAGS: "-D warnings"
    - uses: actions/upload-artifact@v4
      with:
        name: fortitude
        path: target/debug/fortitude

  build-windows:
    name: Build and Test (windows)
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@v4
    - name: "Install cargo nextest"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-nextest
    - name: "Install cargo insta"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-insta
    - name: "Build"
      run: cargo build --verbose
    - name: "Run tests"
      shell: bash
      env:
        NEXTEST_PROFILE: "ci"
      run: cargo insta test --all-features --unreferenced reject --test-runner nextest

  build-macos:
    name: Build and Test (macos)
    runs-on: macos-latest
    steps:
    - uses: actions/checkout@v4
    - name: "Install cargo nextest"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-nextest
    - name: "Install cargo insta"
      uses: taiki-e/install-action@v2.49.35
      with:
        tool: cargo-insta
    - name: "Build"
      run: cargo build --verbose
    - name: "Run tests"
      shell: bash
      env:
        NEXTEST_PROFILE: "ci"
      run: cargo insta test --all-features --unreferenced reject --test-runner nextest

  cargo-install:
    name: Test cargo installation
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: Check cargo install
      run: |
        cargo install --path fortitude
        fortitude --help

  pip-install:
    name: Test pip installation
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - name: Set up Python
      uses: actions/setup-python@v5
      with:
        python-version: '3.x'
    - name: Check pip install
      run: |
        pip install .
        fortitude --help
