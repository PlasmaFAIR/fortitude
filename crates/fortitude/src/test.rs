use std::path::Path;

use anyhow::Result;
use ruff_source_file::{SourceFile, SourceFileBuilder};

use crate::{
    check::{
        ast_entrypoint_map, check_file, read_to_string, rules_to_path_rules, rules_to_text_rules,
    },
    message::{Emitter, TextEmitter},
    rule_table::RuleTable,
    rules::Rule,
    settings::{self, FixMode, Settings},
};

#[macro_export]
macro_rules! apply_common_filters {
    {} => {
        let mut settings = insta::Settings::clone_current();
        // Convert windows paths to Unix Paths.
        settings.add_filter(r"\\\\?([\w\d.])", "/$1");
        let _bound = settings.bind_to_scope();
    }
}

pub(crate) fn test_resource_path(path: impl AsRef<Path>) -> std::path::PathBuf {
    Path::new("./resources/test/").join(path)
}

/// Run [`check_path`] on a Fortran file in the `resources/test/fixtures` directory.
pub(crate) fn test_path(
    path: impl AsRef<Path>,
    rules: &[Rule],
    settings: &Settings,
) -> Result<String> {
    let path = test_resource_path("fixtures").join(path);
    let source = read_to_string(&path)?;
    let filename = path.to_string_lossy();
    let file = SourceFileBuilder::new(filename.as_ref(), source.as_str()).finish();
    Ok(test_contents(&path, &file, rules, settings))
}

pub(crate) fn test_contents(
    path: &Path,
    file: &SourceFile,
    rules: &[Rule],
    settings: &Settings,
) -> String {
    let rule_table = RuleTable::from_iter(rules.iter().cloned());
    let path_rules = rules_to_path_rules(&rule_table);
    let text_rules = rules_to_text_rules(&rule_table);
    let ast_entrypoints = ast_entrypoint_map(&rule_table);

    match check_file(
        &rule_table,
        &path_rules,
        &text_rules,
        &ast_entrypoints,
        path,
        file,
        settings,
        FixMode::Generate,
        settings::IgnoreAllowComments::Disabled,
    ) {
        Ok(violations) => {
            if violations.messages.is_empty() {
                return String::new();
            }

            let mut output = Vec::new();

            TextEmitter::default()
                .with_show_fix_status(true)
                .with_show_fix_diff(true)
                .with_show_source(true)
                .with_unsafe_fixes(crate::settings::UnsafeFixes::Enabled)
                .emit(&mut output, &violations.messages)
                .unwrap();

            String::from_utf8(output).unwrap()
        }
        Err(msg) => {
            panic!("Failed to process: {msg}");
        }
    }
}
