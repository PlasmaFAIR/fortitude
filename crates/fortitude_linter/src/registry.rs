use fortitude_macros::RuleNamespace;
use rayon::str;
use std::str::FromStr; // Needed by strum_macros

pub use crate::rules::Rule;
pub use rule_set::{RuleSet, RuleSetIterator};

mod rule_set;

// Rule categories and identity codes
// ----------------------------------
// Helps to sort rules into logical categories, and acts as a unique tag with which
// users can switch rules on and off.

pub trait AsRule {
    fn rule(&self) -> Rule;
}

impl Rule {
    pub fn from_code(code: &str) -> Result<Self, FromCodeError> {
        let (category, code) = Category::parse_code(code).ok_or(FromCodeError::Unknown)?;
        category
            .all_rules()
            .find(|rule| rule.noqa_code().suffix() == code)
            .ok_or(FromCodeError::Unknown)
    }
}

#[derive(thiserror::Error, Debug)]
pub enum FromCodeError {
    #[error("unknown rule code")]
    Unknown,
}

/// The category of each rule defines the sort of problem it intends to solve.
#[derive(
    Debug,
    PartialEq,
    Eq,
    Clone,
    Hash,
    PartialOrd,
    Ord,
    strum_macros::AsRefStr,
    strum_macros::Display,
    strum_macros::EnumIter,
    strum_macros::EnumString,
    strum_macros::IntoStaticStr,
    RuleNamespace,
)]
#[repr(u16)]
#[strum(serialize_all = "kebab-case")]
pub enum Category {
    /// Failure to parse a file.
    #[prefix = "E"]
    Error,
    /// Detect code that is bug-prone or likely to be incorrect.
    #[prefix = "C"]
    Correctness,
    /// Obsolescent language features, as determined by the Fortran standard.
    #[prefix = "OB"]
    Obsolescent,
    /// Update to modern Fortran features. Used for less severe issues than `Obsolescent`,
    /// and goes beyond recommendations in the Fortran standard.
    #[prefix = "MOD"]
    Modernisation,
    /// Violation of style conventions.
    #[prefix = "S"]
    Style,
    /// Avoid platform/compiler-specific features.
    #[prefix = "PORT"]
    Portability,
    /// Fortitude specific rules.
    #[prefix = "FORT"]
    Fortitude,
}

pub trait RuleNamespace: Sized {
    /// Returns the prefix that every single code that fortitude uses to identify
    /// rules from this category starts with.
    fn common_prefix(&self) -> &'static str;

    /// Returns the variant corresponding to the given prefix.
    fn from_prefix(preifx: &str) -> Result<Self, &'static str>;

    /// Attempts to parse the given rule code. If the prefix is recognized
    /// returns the respective variant along with the code with the common
    /// prefix stripped.
    fn parse_code(code: &str) -> Option<(Self, &str)>;

    #[allow(dead_code)]
    fn name(&self) -> &'static str;

    #[allow(dead_code)]
    fn description(&self) -> &'static str;
}

pub mod clap_completion {
    use clap::builder::{PossibleValue, TypedValueParser, ValueParserFactory};
    use strum::IntoEnumIterator;

    use crate::registry::Rule;

    #[derive(Clone)]
    pub struct RuleParser;

    impl ValueParserFactory for Rule {
        type Parser = RuleParser;

        fn value_parser() -> Self::Parser {
            RuleParser
        }
    }

    impl TypedValueParser for RuleParser {
        type Value = Rule;

        fn parse_ref(
            &self,
            cmd: &clap::Command,
            arg: Option<&clap::Arg>,
            value: &std::ffi::OsStr,
        ) -> Result<Self::Value, clap::Error> {
            let value = value
                .to_str()
                .ok_or_else(|| clap::Error::new(clap::error::ErrorKind::InvalidUtf8))?;

            Rule::from_code(value).map_err(|_| {
                let mut error =
                    clap::Error::new(clap::error::ErrorKind::ValueValidation).with_cmd(cmd);
                if let Some(arg) = arg {
                    error.insert(
                        clap::error::ContextKind::InvalidArg,
                        clap::error::ContextValue::String(arg.to_string()),
                    );
                }
                error.insert(
                    clap::error::ContextKind::InvalidValue,
                    clap::error::ContextValue::String(value.to_string()),
                );
                error
            })
        }

        fn possible_values(&self) -> Option<Box<dyn Iterator<Item = PossibleValue> + '_>> {
            Some(Box::new(Rule::iter().map(|rule| {
                let name = rule.noqa_code().to_string();
                let help = rule.as_ref().to_string();
                PossibleValue::new(name).help(help)
            })))
        }
    }
}

#[cfg(test)]
mod tests {
    use std::mem::size_of;

    use strum::IntoEnumIterator;

    use super::{Category, Rule, RuleNamespace};

    #[test]
    fn documentation() {
        for rule in Rule::iter() {
            assert!(
                rule.explanation().is_some(),
                "Rule {} is missing documentation",
                rule.as_ref()
            );
        }
    }

    #[test]
    fn check_code_serialization() {
        for rule in Rule::iter() {
            assert!(
                Rule::from_code(&format!("{}", rule.noqa_code())).is_ok(),
                "{rule:?} could not be round-trip serialized."
            );
        }
    }

    #[test]
    fn category_parse_code() {
        for rule in Rule::iter() {
            let code = format!("{}", rule.noqa_code());
            let (category, rest) =
                Category::parse_code(&code).unwrap_or_else(|| panic!("couldn't parse {code:?}"));
            assert_eq!(code, format!("{}{rest}", category.common_prefix()));
        }
    }

    #[test]
    fn rule_size() {
        assert_eq!(2, size_of::<Rule>());
    }
}
