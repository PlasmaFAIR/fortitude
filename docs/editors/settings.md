# Settings

The Fortitude Language Server provides a set of configuration options to customise its behaviour along
with the ability to use an existing `pyproject.toml` or `fortitude.toml` file to configure the
linter. This is done by providing these settings while initialising the server.  VS Code provides a
UI to configure these settings, while other editors may require manual configuration. The
[setup](./setup.md) section provides instructions on where to place these settings as per the editor.

## Top-level

### `configuration`

The `configuration` setting allows you to configure editor-specific Fortitude behavior. This can be done
in one of the following ways:

1. **Configuration file path:** Specify the path to a `fortitude.toml` or `pyproject.toml` file that
    contains the configuration. User home directory and environment variables will be expanded.
1. **Inline JSON configuration:** Directly provide the configuration as a JSON object.

The default behavior, if `configuration` is unset, is to load the settings from the project's
configuration (a `fortitude.toml` or `pyproject.toml` in the project's directory), consistent with when
running Fortitude on the command-line.

The [`configurationPreference`](#configurationpreference) setting controls the precedence if both an
editor-provided configuration (`configuration`) and a project level configuration file are present.

#### Resolution order {: #configuration_resolution_order }

In an editor, Fortitude supports three sources of configuration, prioritised as follows (from highest to
lowest):

1. **Specific settings:** Individual settings like [`lineLength`](#linelength) or
    [`lint.select`](#select) defined in the editor
1. [**`fortitude.configuration`**](#configuration): Settings provided via the
    [`configuration`](#configuration) field (either a path to a configuration file or an inline
    configuration object)
1. **Configuration file:** Settings defined in a `fortitude.toml` or `pyproject.toml` file in the
    project's directory (if present)

For example, if the line length is specified in all three sources, Fortitude will use the value from the
[`lineLength`](#linelength) setting.

**Default value**: `null`

**Type**: `string`

**Example usage**:

_Using configuration file path:_

=== "VS Code"

    ```json
    {
        "fortitude.configuration": "~/path/to/fortitude.toml"
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          configuration = "~/path/to/fortitude.toml"
        }
      }
    }
    ```

_Using inline configuration:_

=== "VS Code"

    ```json
    {
        "fortitude.configuration": {
            "check": {
                "extend-select": ["style"],
                "exit-unlabelled-loops": {
                    "allow-unnested-loops": true
                }
            },
        }
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          configuration = {
            check = {
              ["extend-select"] = {"style"},
              ["exit-unlabelled-loops"] = "true"
            },
          }
        }
      }
    }
    ```

### `configurationPreference`

The strategy to use when resolving settings across VS Code and the filesystem. By default, editor
configuration is prioritized over `fortitude.toml` and `pyproject.toml` files.

- `"editorFirst"`: Editor settings take priority over configuration files present in the workspace.
- `"filesystemFirst"`: Configuration files present in the workspace takes priority over editor
    settings.
- `"editorOnly"`: Ignore configuration files entirely i.e., only use editor settings.

**Default value**: `"editorFirst"`

**Type**: `"editorFirst" | "filesystemFirst" | "editorOnly"`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.configurationPreference": "filesystemFirst"
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          configurationPreference = "filesystemFirst"
        }
      }
    }
    ```

### `exclude`

A list of file patterns to exclude from linting. See [the documentation](../settings.md#check_exclude)
for more details.

**Default value**: `null`

**Type**: `string[]`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.exclude": ["**/tests/**"]
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          exclude = ["**/tests/**"]
        }
      }
    }
    ```

### `lineLength`

The line length to use for the linter.

**Default value**: `null`

**Type**: `int`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.lineLength": 100
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          lineLength = 100
        }
      }
    }
    ```

### `fixAll`

Whether to register the server as capable of handling `source.fixAll` code actions.

**Default value**: `true`

**Type**: `bool`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.fixAll": false
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          fixAll = false
        }
      }
    }
    ```

### `logLevel`

The log level to use for the server.

**Default value**: `"info"`

**Type**: `"trace" | "debug" | "info" | "warn" | "error"`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.logLevel": "debug"
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          logLevel = "debug"
        }
      }
    }
    ```

### `logFile`

Path to the log file to use for the server.

If not set, logs will be written to stderr.

**Default value**: `null`

**Type**: `string`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.logFile": "~/path/to/fortitude.log"
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          logFile = "~/path/to/fortitude.log"
        }
      }
    }
    ```

## `codeAction`

Enable or disable code actions provided by the server.

### `fixViolation.enable`

Whether to display Quick Fix actions to autofix violations.

**Default value**: `true`

**Type**: `bool`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.codeAction.fixViolation.enable": false
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          codeAction = {
            fixViolation = {
              enable = false
            }
          }
        }
      }
    }
    ```

## `check`

Settings specific to the Fortitude linter.

### `preview` {: #check_preview }

Whether to enable Fortitude's preview mode when linting.

**Default value**: `null`

**Type**: `bool`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.check.preview": true
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          check = {
            preview = true
          }
        }
      }
    }
    ```

### `select`

Rules to enable by default. See [the documentation](https://docs.astral.sh/fortitude/settings/#check_select).

**Default value**: `null`

**Type**: `string[]`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.check.select": ["E", "F"]
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          check = {
            select = {"E", "F"}
          }
        }
      }
    }
    ```

### `extendSelect`

Rules to enable in addition to those in [`check.select`](#select).

**Default value**: `null`

**Type**: `string[]`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.check.extendSelect": ["W"]
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          check = {
            extendSelect = {"W"}
          }
        }
      }
    }
    ```

### `ignore`

Rules to disable by default. See [the documentation](https://docs.astral.sh/fortitude/settings/#check_ignore).

**Default value**: `null`

**Type**: `string[]`

**Example usage**:

=== "VS Code"

    ```json
    {
        "fortitude.check.ignore": ["E4", "E7"]
    }
    ```

=== "Neovim"

    ```lua
    require('lspconfig').fortitude.setup {
      init_options = {
        settings = {
          check = {
            ignore = {"E4", "E7"}
          }
        }
      }
    }
    ```

## VS Code specific

Additionally, the Fortitude extension provides the following settings specific to VS Code. These settings
are not used by the language server and are only relevant to the extension.

### `enable`

Whether to enable the Fortitude extension. Modifying this setting requires restarting VS Code to take effect.

**Default value**: `true`

**Type**: `bool`

**Example usage**:

```json
{
    "fortitude.enable": false
}
```

### `importStrategy`

Strategy for loading the `fortitude` executable.

- `fromEnvironment` finds Fortitude in the environment, falling back to the bundled version
- `useBundled` uses the version bundled with the extension

**Default value**: `"fromEnvironment"`

**Type**: `"fromEnvironment" | "useBundled"`

**Example usage**:

```json
{
    "fortitude.importStrategy": "useBundled"
}
```

### `interpreter`

A list of paths to Python interpreters. Even though this is a list, only the first interpreter is
used.

This setting depends on the [`fortitude.nativeServer`](#nativeserver) setting:

- If using the native server, the interpreter is used to find the `fortitude` executable when
    [`fortitude.importStrategy`](#importstrategy) is set to `fromEnvironment`.
- Otherwise, the interpreter is used to run the `fortitude-lsp` server.

**Default value**: `[]`

**Type**: `string[]`

**Example usage**:

```json
{
    "fortitude.interpreter": ["/home/user/.local/bin/python"]
}
```

### `path`

A list of path to `fortitude` executables.

The first executable in the list which is exists is used. This setting takes precedence over the
[`fortitude.importStrategy`](#importstrategy) setting.

**Default value**: `[]`

**Type**: `string[]`

**Example usage**:

```json
{
    "fortitude.path": ["/home/user/.local/bin/fortitude"]
}
```

### `trace.server`

The trace level for the language server. Refer to the [LSP
specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#traceValue)
for more information.

**Default value**: `"off"`

**Type**: `"off" | "messages" | "verbose"`

**Example usage**:

```json
{
    "fortitude.trace.server": "messages"
}
```
