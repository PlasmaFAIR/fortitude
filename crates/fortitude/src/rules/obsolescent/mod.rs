pub mod common_blocks;
pub mod computed_goto;
pub mod deprecated_character_syntax;
pub mod entry_statement;
pub mod pause_statement;
pub mod specific_names;
pub mod statement_functions;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::settings::Settings;
    use crate::test::test_path;

    #[test_case(Rule::CommonBlock, Path::new("OB011.f90"))]
    #[test_case(Rule::EntryStatement, Path::new("OB021.f90"))]
    #[test_case(Rule::SpecificName, Path::new("OB031.f90"))]
    #[test_case(Rule::ComputedGoTo, Path::new("OB041.f90"))]
    #[test_case(Rule::PauseStatement, Path::new("OB051.f90"))]
    #[test_case(Rule::DeprecatedCharacterSyntax, Path::new("OB061.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("obsolescent").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
