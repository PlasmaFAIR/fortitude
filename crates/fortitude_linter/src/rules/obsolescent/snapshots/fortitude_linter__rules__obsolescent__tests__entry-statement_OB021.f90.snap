---
source: fortitude/src/rules/obsolescent/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/obsolescent/OB021.f90:7:8: OB021 entry statements are obsolescent, use module procedures with generic interface
  |
5 |        RETURN
6 |
7 |        ENTRY SUB2(A, B, C)
  |        ^^^^^^^^^^^^^^^^^^^ OB021
8 |        RETURN
  |

./resources/test/fixtures/obsolescent/OB021.f90:10:8: OB021 entry statements are obsolescent, use module procedures with generic interface
   |
 8 |        RETURN
 9 |
10 |        ENTRY SUB3
   |        ^^^^^^^^^^ OB021
11 |        RETURN
12 |        END
   |
